### ISSUE TITLE
How to use Dagger2 in library project?

### ISSUE THREAD

**Issue by zeshaoaaa** [google/dagger#481]
Our projects (phone, tablet) need share presenters and components in library project, but Application class in application project, so the DaggerApplicationComponent cannot generate. have any solution? Thanks

**Comment by trevjonez** [comment]
If you have items in the library that depend on details of the component you can define a type of base interface but not annotate it. Then in the application code create another component and annotate that.

so you have in the library:
```java
/** some javadoc explaining how to extend this and what modules to include */
public interface AppComponentBase { ... }
```

and in the application you have:
```java
@Component(modules = {...})
public interface AppComponent extends AppComponentBase {...} 
```

By doing this you defer the generation of the component until the application is being built and dagger has the full graph available as a result.

**Comment by GarimaMathur07** [comment]
I have implemented dagger 2 in my library project and when i am using this library in my main project then i am facing cannot resolve symbol DaggerApplicationComponent issue. 

I tried the above solution but it didn't work for me. Can you please explain in brief. cc @zeshaoaaa @trevjonez

**Comment by kennykaye** [comment]
Experiencing something similar to @GarimaMathur07's issue...
> I have implemented dagger 2 in my library project and when i am using this library in my main project then i am facing cannot resolve symbol DaggerApplicationComponent issue.

I've created a [trivial android app ](https://github.com/GreatCall-KayeK/dagger2-demo)to demo this issue. Interestingly enough, Gradle is perfectly capable of processing the annotations and adding the generated files to the class path. It seems in my case that Android Studio fails to index those generated files, resulting in the "Cannot resolve symbol" error.

I've submitted a ticket to Android Studio which can be tracked [here](https://code.google.com/p/android/issues/detail?id=233191).
